source/common/common/assert.cc

```diff
@@ -28,15 +28,66 @@ class ActionRegistrationImpl : public ActionRegistration {
   static std::function<void()> debug_assertion_failure_record_action_;
 };
 
+class EnvoyBugRegistrationImpl : public ActionRegistration {
+public:
+  EnvoyBugRegistrationImpl(std::function<void()> action) {
+    ASSERT(envoy_bug_failure_record_action_ == nullptr,
+           "An ENVOY_BUG action was already set. Currently only a single action is supported.");
+    count_ = 0;
+    envoy_bug_failure_record_action_ = action;
+  }
+
+  ~EnvoyBugRegistrationImpl() override {
+    ASSERT(envoy_bug_failure_record_action_ != nullptr);
+    envoy_bug_failure_record_action_ = nullptr;
+  }
+
+  static bool shouldLogAndInvoke() {
+    ++count_;
+    // Check if count_ is power of two by its bitwise representation.
+    if ((count_ & (count_ - 1)) == 0) {
```

Comment:
Good point...  Something like this should work:

auto counter_value = ++count_;
if ((counter_value & (counter_value - 1)) == 0) {

Reply:
Acquiring and releasing a mutex is cheaper than a cache miss. If the lock contention on this map is high enough that we notice, I think we have way bigger problems. Even if we _did_ experience high lock contention here that we wanted to do something about, we'd likely be better off building a high-concurrency hash table from many normal hash tables, each with its own [reader-writer] mutex, than to build one lock-free hash table using atomic operations.

Atomic operations become haunted graveyards very, very quickly. They're harder to understand and harder to maintain than code written with higher-level thread-safety constructs.

Reply:
I have a working version like above, but resetting the counters via a map clearing method doesn't have an effect when i have static initialization like "_static_ AtomicCounterPtr ptr_for_file_line".

Is it possible to avoid map lookup (the most expensive mutex locked operation) for each iterations while still allowing a working map reset? Former solved by statics, latter nullifies the statics solution.

Reply:
> because resetting the counters via a map clearing method doesn't have an effect when i have static initialization like "_static_ AtomicCounterPtr ptr_for_file_line".

Reset would need to iterate through the map and set all counters to zero, not erase anything in the map.

I agree with @ahedberg though that if we get to the point that we're contending on this mutex, we have big problems. I don't think having a static is worth it.

Reply:
Use of a mutex and map makes me uncomfortable.  Resetting counters to 0 between does not seem like a realistic goal.

Reply:
I'd be ok with either (static, or map+mutex). I don't feel that strongly about it. I agree that the need for reset can be removed by adding a flag (used only by tests) to not do backoff and take the action on every failure of the condition.

If we do the static std::atomic, I'd like to keep the macro-generated code as small as possible, which can be done by passing the atomic as an argument to `shouldLogAndInvokeEnvoyBugForEnvoyBugMacroUseOnly` instead of the file/line number.

Reply:
Alright -- the current is my map+mutex implementation, i moved the mutex lock to just over the map access.

I have a version fully implemented with static atomics and counter resets (this was easier than a flag, when dealing with parametrized tests). It does indeed pass the atomic to the `shouldLogAndInvokeEnvoyBugForEnvoyBugMacroUseOnly`. Should I push that just for review? Not sure.